### PR TITLE
1.2.0 - Add operation handler

### LINKED ISSUE #3: Add transaction factory
Add a factory to create a transaction based on given operations.

```php
<?php
interface TransactionFactoryInterface
{
    public function create(
        OperationInterface ...$operations
    ): TransactionInterface;
}
```

This allows for factories to decide whether or not they can provide a [shadow transaction](https://github.com/johmanx10/transaction/issues/2). E.g.: a factory for filesystem transactions could first apply the operations to a virtual filesystem.

### PR DESCRIPTION
**Pull request by janmartenjongerius**
The operation handler is better suited in a service oriented application.
It allows to prepare visitors separate from the invoking code and thus separates
concerns about operations and their visitors.

This can be used instead of a Transaction, directly. By invoking it, it will create a transaction on-the-fly and pass along pre-configured visitors.

```php
<?php
use Johmanx10\Transaction\OperationHandler;
use Johmanx10\Transaction\OperationInterface;
use Johmanx10\Transaction\Visitor\LogOperationVisitor;
use Psr\Log\LoggerInterface;

$handler = new OperationHandler();
$handler->attachVisitor(
    /** @var LoggerInterface $logger */
    new LogOperationVisitor($logger)
);

/** @var OperationInterface[] $operations */
$handler->handle(...$operations);
```

The very goal is that the calling code only needs to know about operations and their handler. Not about visitors. ~[I have some more ideas on how we can clean up the calling code.](https://github.com/johmanx10/transaction/pull/5#issuecomment-600892942) They can be added through a separate pull request.~

It's probably easiest to grasp the pull request by looking at the new example file: https://github.com/johmanx10/transaction/blob/7b760791b4a0435699e16899e6d3f056cb009e31/examples/operation-handler

Closes #3

### CODE REVIEW

**Comment by janmartenjongerius** [comment]
~@ashokadewit I have been thinking about two more ways to simplify this library for the end-user.

1. Create a formatter that accepts any type of exception, but knows to apply a special formatter for certain library specific exceptions. This would simplify the `catch` statements and bodies for the calling code.
2. Make it so the operation handler uses that internally and instead of returning `void`, it will return a result object which exposes whether the transaction succeeded and holds the exception as well as the formatted exception. - Conversely it could throw a more generic `OperationException` which holds the previous exception like any exception does, plus the specifically formatted exception message with the aforementioned formatter.

If either or both of those features seem interesting to you, I will add those to this release as well. My personal preference goes to both *1* and the Exception implementation of *2*. However, I could very well get away with implementing *2* without implementing *1*, because it would mean I have to create 2 catch clauses for internal exceptions. That's doable, with the 2 formatters that go along with that. Just less pretty and less useful to an end-user trying to implement it differently.~

# Implementation

```php
<?php
public function handle(OperationInterface ...$operations): void
{
    $transaction = new Transaction(...$operations);

    try {
        $transaction->commit(...array_values($this->visitors));
    } catch (Throwable $exception) {
        throw new OperationException(
            $this->exceptionFormatter->format($exception),
            $exception
        );
    }
}
```

# Calling code

```php
<?php
try {
    $handler->handle(...$operations);
} catch (OperationException $exception) {
    // Get the formatted message.
    $exception->getMessage();
    // Get the rollback or failed rollback exception.
    $exception->getPrevious();
}
```

**Comment by janmartenjongerius** [comment]
It seems this closes #3 , as it replaces the need for a factory in a service oriented application.

File: src/OperationHandler.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    /**
+     * Handle the given operations.
+     *
+     * @param OperationInterface ...$operations
+     *
+     * @return void
+     */
+    public function handle(OperationInterface ...$operations): void
+    {
+        $transaction = new Transaction(...$operations);
```
Comment (janmartenjongerius) [inline src/OperationHandler.php]: I could decouple this by introducing a factory for transactions and optionally injecting that in the constructor of the operation handler. Would you prefer this, @ashokadewit ?

**Comment by janmartenjongerius** [comment]
After having a look at the available formatters and exceptions, seeing their interfaces aren't able to share a parent interface before PHP 7.4, I have come up with a different approach.

I could inject a factory for creating an operation exception, based on any throwable, into the operation handler. That way, the factory for this exception holds shared logic to format the exception messages, but additionally it can determine the affected operation and pass that along to the operation exception, to further simplify exceptions when debugging failed operations.

I would like the time to make this part of the current pull request.

**Review by ashokadewit — APPROVED** [review]

**Pull request merged**